### ISSUE TITLE
Support for custom service calls

### ISSUE THREAD

**Issue by CountParadox** [PRProd/HA-Firemote#152]
**Describe the solution you'd like**

I would like to be able to have 'app shortcut's call custom services. This way I could call a service that would be able to change the inputs on my AV receiver between for example. my chromecast and my kodi box, or my nintendo switch.

I currently have buttons on my dashboard that call these services, it would look nicer if they were part of my 'virtual remote'

**Comment by PRProd** [comment]
I like that idea.  It will take some work, but I think it would be quite helpful for a lot of people.  Let me think on this for a while and see what I can come up with.

**Comment by zhdenny** [comment]
This would be GREAT! One of the extra app lunch buttons could then trigger any Home Assistant Service Call :)

**Comment by mattague** [comment]
This would also be useful for the volume buttons. I have an av receiver for my audio but the buttons only control the shield

**Comment by PRProd** [comment]
@mattague - If you are only interested in re-targeting your Firemote volume buttons from your Shield to your receiver, you can do that today by using [button overrides](https://github.com/PRProd/HA-Firemote#button-overrides)

**Comment by mattague** [comment]
I actually realized I could just setup CEC properly, but thank you for replying.

**Comment by PRProd** [comment]
Great news @CountParadox!

I have just published Firemote beta release version [v3.1.0.b1](https://github.com/PRProd/HA-Firemote/releases/tag/v3.1.0b1) for you to download and test!

This new feature will be accessed using YAML in one of two ways:  Define a script or Construct a Service Call
```yaml
type: custom:firemote-card
entity: media_player.fire_tv_192_168_107_88
device_family: amazon-fire
device_type: fire_tv_4_series
compatibility_mode: default
custom_launchers:
  - friendly_name: flash the light
    label: flash
    script: flashthelablight
  - friendly_name: Red Lamp
    label: red
    service: light.toggle
    target:
      entity_id: light.bedroom_lamp
    data:
      color_name: red
      transition: 2
      brightness_pct: 100
app_launch_1: customlauncher_1
app_launch_2: customlauncher_2
app_launch_3: youtube
```
<br><br>
Firstly, to define a custom launcher, you will need to open with a **custom_launchers:** key in your YAML.  Next, you will need to define your custom launcher button using the values in the table below:
<br><br>
| Key | Value(s) | Required | Description |
| ---- | ------------| ------------- | ---------------|
| friendly_name | string  | yes | This is the name that will appear in the GUI app selector dropdown |
| label                | string  | no  | The text that will appear on the button |
| script               | string  | yes if calling a script | The name of your Home Assistant script |
| service            | string  | yes if not using a script | The name of the Home Assistant service you want to call |
| target              | object | yes if not using a script | Define the entity for your service to act upon |
| data                | object | no | optional parameters for your service call |

<br><br>
Next, you can use your custom buttons by assigning them through YAML with the name **customlauncher_n** (where n is the number of the launcher you defined), or you can select them by the Friendly Name you gave them by using the Visual Editor:

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/d3e4a3d5-8937-4908-9704-7e63d7f0b365)

<br><br>
Please let me know if my instructions are unclear, if it doesn't work, or if it works perfectly.  I'd really love to have your feedback before making this new feature a part of the official release.

**Comment by CountParadox** [comment]
Looks awesome!


In the yaml,

You wrote:
app_launch_1: customlauncher_1

Would this be the first item in the list? Or would this be one with friendly name "customlauncher 1" ?

I was not sure about this but I can see in the UI how it works !

**Comment by PRProd** [comment]
@CountParadox - You are correct.  _customlauncher\_1_ refers to the first "child" in your YAML under the **custom_launchers:** key.  

If you had 10 **custom_launchers** defined in your YAML, and you wanted to put your third defined launcher into Firemote button 4 you'd use
```yaml
app_launch_4: customlauncher_3
```

It's a little clunky, I know.  It's a lot less clunky if you're willing to bounce back over to the Visual Editor and pick it from there.  Is it too confusing, or do you think that I'd be fine to keep it this way?

**Comment by CountParadox** [comment]
I understand it, and it is what I thought, but it is different to how homeassistant itself handles things,
Would it be difficult to replicate the home assistant / esphome method of giving each one an ID and then using that?


Eg. 


  - platform: binary
    name: "TV Lamp"
    Id: light_output

**Comment by PRProd** [comment]
It would have been easier if I'd have predicted this request before I began building the structure of Firemote.  I kind of backed myself into a corner which is how I ended up with this awkward solution.

I'm going to think about it some more to see if I can come up with a better idea that makes more sense for the YAML side.

**Comment by Jaykurb** [comment]
Are we able to use icons instead of text for the custom buttons?

**Comment by PRProd** [comment]
@Jaykurb - Not yet.  What kind of icons are you interested in using?  Did you want to use your own SVG artwork, or MDI icons from Home Assistant, or ... something else?

**Comment by Jaykurb** [comment]
I was thinking of mdi icons but artwork would be cool.

**Comment by PRProd** [comment]
## BREAKING CHANGE
@CountParadox ~ Good news!  I realized that I wasn't as stuck as I thought I was with the YAML naming thing.  I just published beta release version [v3.1.0.b2](https://github.com/PRProd/HA-Firemote/releases/tag/v3.1.0b2) which uses this NEW structure:
```yaml
type: custom:firemote-card
entity: media_player.fire_tv_192_168_107_88
device_family: amazon-fire
device_type: fire_tv_4_series
compatibility_mode: default
custom_launchers:
  - friendly_name: flash the light
    label: flash
    script: flashthelablight
  - friendly_name: Red Lamp
    label: red
    service: light.toggle
    target:
      entity_id: light.bedroom_lamp
    data:
      color_name: red
      transition: 2
      brightness_pct: 100
app_launch_1: customlauncher Red Lamp
app_launch_2: customlauncher flash the light
app_launch_3: youtube
```

Everyone - Take note that the **app_launch_n** YAML keys now have values that are **customlauncher** -space- and the friendly_name of your custom launcher

**Comment by CountParadox** [comment]
I like this more, it's easier to follow :) 

That said I'll still use the UI....

**Comment by PRProd** [comment]
Thanks.  I feel better about it too.  :smile:

Have you tested it yet?  I haven't heard anybody mention anything good or bad about the new functionality yet.

**Comment by PRProd** [comment]
I found a small UI bug that only appeared in the iOS version of the HA companion app.  If you notice a button background that's larger than the button should be, that fix will be included in the next release.

**Comment by PRProd** [comment]
I have just published Firemote version [v3.1.0](https://github.com/PRProd/HA-Firemote/releases/tag/v3.1.0) which officially closes this request!

Thanks everybody for your input and testing.
